export.go

```diff
@@ -3,65 +3,186 @@ package containerd
 import (
 	"archive/tar"
 	"context"
+	"encoding/json"
 	"io"
+	"sort"
 
 	"github.com/containerd/containerd/content"
 	"github.com/containerd/containerd/images"
-	"github.com/containerd/containerd/oci"
 	ocispecs "github.com/opencontainers/image-spec/specs-go"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
 )
 
 func (c *Client) exportToOCITar(ctx context.Context, desc ocispec.Descriptor, writer io.Writer, eopts exportOpts) error {
 	tw := tar.NewWriter(writer)
-	img := oci.Tar(tw)
+	defer tw.Close()
 
-	// For tar, we defer creating index until end of the function.
-	if err := oci.Init(img, oci.InitOpts{SkipCreateIndex: true}); err != nil {
-		return err
+	records := []tarRecord{
+		ociLayoutFile(""),
+		ociIndexRecord(desc),
 	}
+
 	cs := c.ContentStore()
+	algorithms := map[string]struct{}{}
+	exportHandler := func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		records = append(records, blobRecord(cs, desc))
+		algorithms[desc.Digest.Algorithm().String()] = struct{}{}
+		return nil, nil
+	}
+
 	handlers := images.Handlers(
 		images.ChildrenHandler(cs),
-		exportHandler(cs, img),
+		images.HandlerFunc(exportHandler),
 	)
-	// For tar, we need to use Walk instead of Dispatch for ensuring sequential write
+
+	// Walk sequentially since the number of fetchs is likely one and doing in
+	// parallel requires locking the export handler
 	if err := images.Walk(ctx, handlers, desc); err != nil {
 		return err
 	}
-	// For tar, we don't use oci.PutManifestDescriptorToIndex() which allows appending desc to existing index.json
-	// but requires img to support random read access so as to read index.json.
-	return oci.WriteIndex(img,
-		ocispec.Index{
-			Versioned: ocispecs.Versioned{
-				SchemaVersion: 2,
-			},
-			Manifests: []ocispec.Descriptor{desc},
+
+	if len(algorithms) > 0 {
+		records = append(records, directoryRecord("blobs/", 0755))
+		for alg := range algorithms {
+			records = append(records, directoryRecord("blobs/"+alg+"/", 0755))
+		}
+	}
+
+	return writeTar(ctx, tw, records)
+}
+
+type tarRecord struct {
+	Header *tar.Header
+	CopyTo func(context.Context, io.Writer) (int64, error)
+}
+
+func blobRecord(cs content.Store, desc ocispec.Descriptor) tarRecord {
+	path := "blobs/" + desc.Digest.Algorithm().String() + "/" + desc.Digest.Hex()
+	return tarRecord{
+		Header: &tar.Header{
+			Name:     path,
+			Mode:     0444,
```

Comment:
I believe it was. The original implementation had these as `0444` and it makes sense considering these files could be considered immutable by their hash reference.